### PR TITLE
Ping before sending FindNode

### PR DESCRIPTION
**Pull request by gumb0**
Addresses point 4 of https://github.com/ethereum/aleth/issues/4959

### CODE REVIEW

File: test/unittests/libp2p/net.cpp [hunk 1]
```diff
@@ -172,10 +199,10 @@ struct TestNodeTable: public NodeTable
     {
         NodeTable::onPacketReceived(_socket, _from, _packet);
 
-        packetReceived.set_value(_packet.toBytes());
```
Comment (gumb0) [inline test/unittests/libp2p/net.cpp]: This was not working correctly when more than one packet was received - `set_value` throws if `promise` already has a value.

**Comment by gumb0** [comment]
The unit test is not great, but it's something. Don't want to spend more time on it now, maybe I'll improve it / add more tests later.

File: test/unittests/libp2p/net.cpp [hunk 2]
```diff
@@ -388,7 +414,7 @@ BOOST_AUTO_TEST_CASE(noteActiveNodeAppendsNewNode)
     TestNodeTableHost nodeTableHost(1);
     nodeTableHost.populate(1);
 
-    auto nodeTable = nodeTableHost.nodeTable;
+    auto& nodeTable = nodeTableHost.nodeTable;
```
Comment (halfalicious) [inline test/unittests/libp2p/net.cpp]: Nice catch :smile:

**Comment by halfalicious** [comment]
As you mentioned, it would be nice to add some tests which verify that the ping is only sent before the FindNode if the pong time is older than 12 hours (this would probably require some changes to `TestNodeTableHost`) but that can be taken care of in a future change.

File: test/unittests/libp2p/net.cpp [hunk 3]
```diff
@@ -56,6 +56,33 @@ class TestHost: public Worker
     ba::io_service m_io;
 };
 
+template <typename T>
+class ConcurrentQueue
```
Comment (chfast) [inline test/unittests/libp2p/net.cpp]: Have you seen https://github.com/ethereum/aleth/blob/master/libdevcore/concurrent_queue.h ?

File: test/unittests/libp2p/net.cpp (hunk 2 above)
Comment (chfast) [reply test/unittests/libp2p/net.cpp]: Maybe block copy constructor?

Comment (gumb0) [reply test/unittests/libp2p/net.cpp]: It's a `shared_ptr` anyway

File: test/unittests/libp2p/net.cpp (hunk 3 above)
Comment (gumb0) [reply test/unittests/libp2p/net.cpp]: Oh I haven't! Looks very similar, I'll remove mine then.

Where is it used though?

Comment (chfast) [reply test/unittests/libp2p/net.cpp]: Seems nowhere currently. I was using it for SmartVM previously to JIT contracts in background thread. You can adjust it to your needs.

**Pull request merged**